test/libevmasm/Optimiser.cpp

```diff
@@ -1467,6 +1478,91 @@ BOOST_AUTO_TEST_CASE(cse_replace_too_large_shift)
 	});
 }
 
+BOOST_AUTO_TEST_CASE(inliner)
+{
+	AssemblyItem jumpInto{Instruction::JUMP};
+	jumpInto.setJumpType(AssemblyItem::JumpType::IntoFunction);
+	AssemblyItem jumpOutOf{Instruction::JUMP};
+	jumpOutOf.setJumpType(AssemblyItem::JumpType::OutOfFunction);
+	AssemblyItems items{
+		AssemblyItem(PushTag, 1),
+		AssemblyItem(PushTag, 2),
+		jumpInto,
+		AssemblyItem(Tag, 1),
+		Instruction::STOP,
+		AssemblyItem(Tag, 2),
+		Instruction::CALLVALUE,
+		Instruction::SWAP1,
+		jumpOutOf,
+	};
+	AssemblyItems expectation{
+		AssemblyItem(PushTag, 1),
+		Instruction::CALLVALUE,
+		Instruction::SWAP1,
+		Instruction::JUMP,
+		AssemblyItem(Tag, 1),
+		Instruction::STOP,
+		AssemblyItem(Tag, 2),
+		Instruction::CALLVALUE,
+		Instruction::SWAP1,
+		jumpOutOf,
```

Comment:
NTS: This is

```
CALLVALUE
SWAP1
JUMP
Tag_1
STOP
Tag_2
CALLVALUE
SWAP1
JUMP
```

Reply:
Does this mean the following?

```
1. Callvalue, SWAP1, and jump to location specified by stack top before callvalue was issued (return addr is value returned by callvalue)
```

Sorry a bit confused how to read this. Dunno if I am reading it correctly in the first place :-)

Reply:
This is the bare "inlined" version of the above.
``PushTag(1)`` still pushes the return point for the following jump. Then ``CALLVALUE`` is the "return value" of the inlined function. ``SWAP1`` rearranges the stack for jumping, s.t. the following jump jumps to ``Tag_1``.
So in both cases you end up at ``STOP`` with the callvalue on stack - i.e. the code is equivalent.

In practice there are more optimizer steps involved. In particular CSE will move the ``PushTag(1)`` down towards the jump, eliminating the SWAP1, i.e. the next step would be:

```
		Instruction::CALLVALUE,
		AssemblyItem(PushTag, 1),
		Instruction::JUMP,
		AssemblyItem(Tag, 1),
		Instruction::STOP,
		AssemblyItem(Tag, 2),
		Instruction::CALLVALUE,
		Instruction::SWAP1,
		jumpOutOf,
```
Then the peephole optimizer removes ``Push(Tag_1) jump tag_1``, i.e. you get
```
		Instruction::CALLVALUE,
		AssemblyItem(Tag, 1),
		Instruction::STOP,
		AssemblyItem(Tag, 2),
		Instruction::CALLVALUE,
		Instruction::SWAP1,
		jumpOutOf,
```
Furthermore ``tag_1`` and  ``tag_2``  would now be unused and be removed and the code behind ``tag_2`` removed, because it's unreachable, yielding:
```
		Instruction::CALLVALUE,
		Instruction::STOP
```
in the end :-).

Most of this has nothing to do with this PR, though.

Reply:
Ah I forgot to read the first push tag instruction in the optimised sequence :)